### ISSUE TITLE
No immutability

### ISSUE THREAD

**Issue by AndreaCrotti** [pytoolz/toolz#280]
I just noticed one issue with dissoc on dictionaries.
```python
diff --git a/toolz/tests/test_dicttoolz.py b/toolz/tests/test_dicttoolz.py
index 9134276..9b9a735 100644
--- a/toolz/tests/test_dicttoolz.py
+++ b/toolz/tests/test_dicttoolz.py
@@ -103,6 +103,10 @@ class TestDict(object):
         assert d is oldd
         assert d2 is not oldd
 
+        d = D({'x': {'a': 1}, 'y': 1})
+        d2 = dissoc(d, 'y')
+        assert id(d['x']) != id(d2['x'])
+
     def test_update_in(self):
         D, kw = self.D, self.kw
         assert update_in(D({"a": 0}), ["a"], inc, **kw) == D({"a": 1})

```

I just added an example test to see that it fails, in short if you have nested dictionaries copy.copy doesn't create completely new and independent objects, but subobjects are shared.
Using copy.deepcopy instead would do the trick (but it's slower obviously)

**Comment by Digenis** [comment]
`deepcopy` is not used anywhere in toolz.
I prefer it as is
because composing functions with deepcopy is easy
but mutating some nested structure is impractical/impossible
after the original references are lost by copying.

I think it's possible to come up with a toolz namespace that deepcopies
but this should be justified by some serious use cases first.

**Comment by AndreaCrotti** [comment]
@Digenis I'm not sure why I should lose the original reference..

I mean if dissoc is functional it should return a new object, not an object that has still references to the object passed in.

Otherwise I could do something like this:

```python
d1 = {'a': {1: 2}, 'b': 2}
d2 = dissoc(d1, 'b')

d2['a'][1] = 10
# at this point also d1['a'][1] is 10
continue_computation(d1)
```

I understand that performances are a bit crap (and we need maybe something like that to improve them https://github.com/tobgu/pyrsistent) in Python with deepcopy, but the above behaviour is not really right to me..

**Comment by eriknw** [comment]
Yeah, `deepcopy` isn't used, because it would be far less efficient.  `toolz.dictoolz` enables functional-style operations on dicts and nested structures.  It allows users to choose not to mutate objects by convention; it does not enforce immutability.  `toolz.dicttoolz` includes several common functions for making changes to a structure, which don't modify the original.

@AndreaCrotti, did this behavior affect code that you wrote?  What manual operation did you do?  Do you think the docs should be more clear about this behavior?

**Comment by AndreaCrotti** [comment]
Well the doc says
```
Return a new dict with the given key(s) removed.

New dict has d[key] deleted for each supplied key.
Does not modify the initial dictionary.
```

which is not entirely true in the sense that the dictionary is not all new if it contains references to the old one.
In the real world maybe I won't have issues often, but I think it could be quite confusing for people coming from real functional languages, where this would never happen (thanks to persistent data structures).
So for me the default behaviour should be to use *deeocopy* and have an option to not use it at your own risk, but well  also the other way around could be acceptable..

**Comment by Digenis** [comment]
It is true because it's a reference to a new dictionary.
It states nothing about its values.

I like to see this ticket as an opinion towards all toolz, not just dissoc.
This is why I brought up the separate, namespace.
I would use such a namespace, probably more than the default one.

**Comment by eriknw** [comment]
What is a motivating use case for doing a deep copy?

Also:
```python
dissoc = lambda *args: copy.deepcopy(toolz.dissoc(*args))
```
or
```python
def fullcopy(func):
    def inner(*args, **kwargs):
        return copy.deepcopy(func(*args, **kwargs))
    return inner

dissoc = fullcopy(toolz.dissoc)
```

**Comment by AndreaCrotti** [comment]
Well @eriknw that would not be great since it would do two copies, making it even more inefficient.

At that point I can just do:
```python
newdict = copy.deepcopy(dic)
del newdict['key']
```
which would still be exactly the same and copying only once as far as I'm aware.

All I'm saying is that this is not what I would have expected, but if you don't want to change the behaviour maybe just an extra comment somewhere in the README or the functions doc might be enough..

**Comment by eriknw** [comment]
We are not going to do deep copies in `toolz.dictoolz`.  When would you ever want to do this?  I still don't see a motivating use case.  `dict.copy`, `set.copy`, etc. are all shallow copies.  Shallow copies should always be expected to be the default in Python.  *I* would be surprised otherwise.  I'm not against making this more clear in the docs, though, so your feedback is appreciated.  Where would be the best place to add a note (i.e., what would have minimized your surprise at this behavior)?

**Comment by AndreaCrotti** [comment]
So well I think if it's done in the same way almost everywhere is not worth
changing the docstring in every function

We could just mention in the Readme that since Python doesn't have
persistent data structures returning deep copies of dicts/sets would be too
expensive.
I think what would be enough, and maybe noone else is really bothered by
this anyway..

**Comment by AlJohri** [comment]
Would there be any interest in using the persistent data structures from  [pyrsistent](https://github.com/tobgu/pyrsistent) or [fn.py](https://github.com/kachayev/fn.py) in a separate namespace?
